rmw_zenoh_cpp/src/detail/graph_cache.cpp

```diff
@@ -270,33 +191,185 @@ static std::vector<std::string> split_keyexpr(const std::string & keyexpr)
+    return std::nullopt;
+  }
+  for (const std::string & p : parts) {
+    if (p.empty()) {
+      RCUTILS_LOG_ERROR_NAMED(
+        "rmw_zenoh_cpp",
+        "Received invalid liveliness token");
+      return std::nullopt;
+    }
   }
+
   // Get the entity, ie N, MP, MS, SS, SC.
-  const std::string & entity = parts[2];
+  std::string & entity = parts[2];
+  if (entity != "NN" &&
+    entity != "MP" &&
+    entity != "MS" &&
+    entity != "SS" &&
+    entity != "SC")
+  {
+    RCUTILS_LOG_ERROR_NAMED(
+      "rmw_zenoh_cpp",
+      "Invalid entity [%s] in liveliness token", entity.c_str());
+    return std::nullopt;
+  }
+
+  GraphNode node;
+  // Nodes with empty namespaces will contain a "_".
+  node.ns = parts.at(3) == "_" ? "/" : "/" + parts.at(3);
+  node.name = std::move(parts[4]);
+
+  if (entity != "NN") {
+    if (parts.size() < 8) {
+      RCUTILS_LOG_ERROR_NAMED(
+        "rmw_zenoh_cpp",
+        "Received invalid liveliness token");
+      return std::nullopt;
+    }
+    GraphNode::TopicData data;
+    data.topic = "/" + std::move(parts[5]);
+    data.type = std::move(parts[6]);
+    data.qos = std::move(parts[7]);
+
+    if (entity == "MP") {
+      node.pubs.push_back(std::move(data));
+    } else if (entity == "MS") {
+      node.subs.push_back(std::move(data));
+    } else {
+      // TODO.
+    }
+  }
+
+  return std::make_pair(std::move(entity), std::move(node));
+}
+} // namespace anonymous
+
+///=============================================================================
+GraphCache::TopicStats::TopicStats(std::size_t pub_count, std::size_t sub_count)
+: pub_count_(pub_count),
+  sub_count_(sub_count)
+{
+  // Do nothing.
+}
+
+///=============================================================================
+void GraphCache::parse_put(const std::string & keyexpr)
+{
+  auto valid_token = _parse_token(keyexpr);
+  if (!valid_token.has_value()) {
+    // Error message has already been logged.
+    return;
+  }
+  const std::string & entity = valid_token->first;
+  auto node = std::make_shared<GraphNode>(std::move(valid_token->second));
   std::lock_guard<std::mutex> lock(graph_mutex_);
+
   if (entity == "NN") {
     // Node
-    RCUTILS_LOG_WARN_NAMED("rmw_zenoh_cpp", "Adding node %s to the graph.", parts.back().c_str());
-    const bool has_namespace = entity.size() == 5 ? true : false;
-    graph_[parts.back()] = YAML::Node();
-    // TODO(Yadunund): Implement enclave support.
-    graph_[parts.back()]["enclave"] = "";
-    graph_[parts.back()]["namespace"] = has_namespace ? parts.at(4) : "/";
+    auto ns_it = graph_.find(node->ns);
+    if (ns_it == graph_.end()) {
+      // New namespace.
+      std::string ns = node->ns;
+      std::unordered_map<std::string, GraphNodePtr> map = {
+        {node->name, node}
+      };
+      graph_.insert(std::make_pair(std::move(ns), std::move(map)));
+    } else {
+      auto insertion = ns_it->second.insert(std::make_pair(node->name, node));
+      if (!insertion.second) {
```

Comment:
We can rely on `std::find_if` from `<algorithm>` here as we are in cpp 17

Reply:
`std::find_if()` will have O(n) complexity while the insertion approach should have `O(log n)`. The idea is that insertion will fail if the entry exists and will give us an iterator to the element that already exists which can be updated.

Reply:
Rust has a nicer way to do this with HashMap's`entry().or_insert()` [pattern](https://stackoverflow.com/questions/30414424/how-can-i-update-a-value-in-a-mutable-hashmap).